problem_three.rb

```diff
@@ -0,0 +1,19 @@
+# Given an array of numbers having size N in which every number is between 1 and N, determine if there are any duplicates in it. 
+# - Do not sort the array or use bit vectors.
+# - Do not use 2 nesting for loops since the time complexity would be high.
+# - You are allowed to destroy the array if you like.
+# - Try to work within the array without using any other temporary data structures.
+# - Try to find a solution by executing just one loop.
+
+class ProblemThree
+  def duplicates?(array)
+    array.each_with_index do |_, index|
+      raise "array#{[index - 1]}=#{array[index - 1]} outside 1 and #{array.length}" \
+        if array[index - 1].abs > array.length
+      return true if array[array[index - 1].abs - 1] < 0
+
+      array[array[index - 1].abs - 1] > 0 && array[array[index - 1].abs - 1] *= -1
```

Comment:
This is not better refactor. It is less code, but remember we are basically writing software mainly humans, not machines. Someone else will have to maintain the code you write and you don't want to make his/her life hard (say you should be maintaining someone else's code)

See https://i2.wp.com/commadot.com/wp-content/uploads/2009/02/wtf.png?w=550